### DISCUSSION TITLE
Show status message when updating command line buffer

### DISCUSSION THREAD

**Discussion by Realiserad** [fish-shell/fish-shell#10380] [Q&A] [answered]
I have made a plugin which replaces the command line buffer using `commandline --replace`. It works, but if the plugin fails for some reason, it would be nice to show a status emoji somewhere (e.g. ❌) to indicate to the user that an error occurred.

Is there such a mechanism in fish?

**Comment by zanchey** [comment]
The right-side prompt might be the place to put it.

**Reply by Realiserad** [reply]
That's what I figured as well! However, I couldn't get it to put anything in [the right-side prompt](https://fishshell.com/docs/current/cmds/fish_right_prompt.html) without rendering the prompt again (i.e. to press *Enter* on the keyboard). Did I miss anything? 🙂

**Reply by zanchey** [reply]
You might need to call `commandline -f repaint` to redraw the right prompt.

(In future, it might be worth including what you have tried already, to improve the quality of responses)

**Reply by Realiserad** [reply]
I made a uhum.. clumsy attempt. 😇

Created a `.config/fish/conf.d/foo.fish` which looks like this:

```shell
bind \n foo

function foo
  # Complicated business logic producing an error every time...
  # Put status emoji in the terminal
  eval (echo "function fish_right_prompt; echo '❌'; end")
  commandline --function repaint
  # Great it works! But how do I get rid of it when the user
  # presses *Enter*?
end
```

Restarted the terminal and pressed Ctrl+J to trigger the `foo` function. Unfortunately the ❌ sticks around after pressing e.g. Ctrl+C or Enter to get a new prompt. I tried to add `functions --erase fish_right_prompt` at the end of the `foo` function but then it doesn't display the ❌ at all.

**Reply by zanchey** [reply, marked as answer]
If you want to wipe it on those keys, you would need something like this:

```
function fish_right_prompt
   echo $mychar
end

function mybinding
   # Complicated business logic producing an error every time...
  set -g mychar ❌
  commandline --function repaint
end

bind \cj mybinding

function clearstatus
  set -e mychar
end

bind \r 'clearstatus; commandline -f execute'
bind \cc 'clearstatus; commandline -f cancel-commandline'
```

**Reply by Realiserad** [reply]
Thank you, @zanchey, I really appreciate the help!

Following your advice, I was able to get it to work (for reference, [here is my commit](https://github.com/Realiserad/fish-ai/commit/73c00d37e5d0173e7a12114bdf5e929873e1ad78)). 

There is only one small issue I cannot wrap my head around. The complicated business logic takes a while to execute, so it would be great to show a ⏳ when it's "thinking". 

I tried to put a `set -g mychar ⏳; commandline -f repaint` in the beginning of the function, but it doesn't do anything. A one-minute digging into Rust suggest that `repaint` puts something on a queue and that the actual repaint happens at a later stage. Is there a way to repaint immediately?

**Reply by zanchey** [reply]
The [documentation for commandline](https://fishshell.com/docs/current/cmds/commandline.html) does spell this out - the arguments to `--function` are added to the queue, but it's not processed until the reader is ready to take more input.
